[flinksink] Add base implementation for writing files to the FS

pkubit-g opened the pull request.

allisonport-db committed:
update package object path (#206)

allisonport-db committed:
Close ParquetIterable in SnapshotImpl.loadInMemory (#207)

scottsand-db commented on flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java:
```diff
@@ -78,20 +79,31 @@
+    /**
+     * This method creates the {@link SinkWriter} instance that will be responsible for passing
+     * incoming stream events to the correct bucket writer and then flushed to the underlying files.
+     * <p>
```
nit: no need for this `<p>` here, unless in future PRs there will be more text below.

scottsand-db commented on flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java:
```diff
@@ -78,20 +79,31 @@
+     * @param context {@link SinkWriter} init context object
+     * @param states  restored states of the writers. Will be empty collection for fresh start.
+     * @return new {@link SinkWriter} object
+     * @throws IOException Thrown, if the recoverable writer cannot be instantiated.
```
nit: don't think we need the extra word `Thrown`

scottsand-db commented on flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java:
```diff
@@ -145,15 +157,50 @@
+
+    /**
+     * Convenience method for creating {@link DeltaSink}.
+     * Developer must ensure that writerFactory is configured to use SNAPPY compression codec.
```
just so you know, when we generate javadocs, newlines are removed and only `<p>` actually provides a new line in the output. So
```
Convenience method for creating {@link DeltaSink}.
Developer must ensure that writerFactory is configured to use SNAPPY compression codec.
```

just becomes

```
Convenience method for creating {@link DeltaSink}. Developer must ensure that writerFactory is configured to use SNAPPY compression codec.
```

scottsand-db commented on flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java:
```diff
@@ -45,32 +59,310 @@
+    ///////////////////////////////////////////////////////////////////////////
+
+    ///////////////////////////////////////////////////////////////////////////
+    // configuration fields
```
if `configuration fields` is a "sub-header" of `FileSink-specific fields`, can you make the `\\\\\\\\...` shorter?
i.e.
```
///////////////////////////////////////////////////////////////////////////
// FileSink-specific fields
///////////////////////////////////////////////////////////////////////////
...
///////////////////////////////////////////////////
// configuration fields
///////////////////////////////////////////////////
```

scottsand-db commented on flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java:
```diff
@@ -57,39 +70,313 @@
+        final Path partFilePath = assembleNewPartPath();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(
```
nit: is there no way to move these `LOG.isDebugEnabled` checks to some common `Logging` class?
i.e. most codes just say `logDebug(...)` and then the Logging class performs the `LOG.isDebugEnabled` check for us.

pkubit-g replied:
I was just following Flink's style of coding here where they are not using any helper methods, but sure we can apply one.
But can you tell what's the concept behind and the full interface of proposed `logDebug` method  ? 
As far as i understand `isDebugEnabled` logic here it exists so as not to materialize the String's objects if they will not be used. But if we move the check and the actual logging to a helper method then we will be creating those objects on every method's call which will impact latency.

scottsand-db replied:
Great point! Then this as it is is fine :)

scottsand-db reviewed (commented):
Did a first scan - will do more. Looks great! Great documentation - really helps with the readability.

allisonport-db committed:
javadocs with return statements' (#210)

allisonport-db committed:
[#124] Fix for getBigDecimal decoding bug with parquet4s LongValue (#208)

Resolves #124 
• Added `Parquet4s.LongValue` to `customDecimalCodec`
• Added test case to `DeltaDataReaderSuite`
• Created table for test in `GoldenTables.scala`

pkubit-g committed:
Flink Delta Sink - PR 1 (#196)

* Add base classes and interfaces for Flink Delta Sink

* Apply code review remarks

pkubit-g committed:
Add base implementation for writing files to the FS

tdas commented on flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java:
```diff
@@ -45,32 +59,310 @@
+     * @param context context for getting additional data about input event
+     * @implNote This method behaves in the same way as
+     * {@link org.apache.flink.connector.file.sink.writer.FileWriter#write}
+     * except that it uses custom {@link DeltaWriterBucket} implementation.
```
add why it uses the custom one

tdas commented on flink-connector/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaBulkBucketWriter.java:
```diff
@@ -0,0 +1,83 @@
+import org.apache.flink.util.Preconditions;
+
+/**
+ * This class is provided as a part of workaround for getting actual file size.
```
not clear what this class does. the reason it exists does not quite make sense without more information abouut what it does

pkubit-g replied:
actually it is explained below:
```
Compared to its original version {@link BulkPartWriter} it changes only the return types
for methods {@link this#resumeFrom} and {@link this#openNew} to a custom implementation
of {@link BulkPartWriter} that is {@link DeltaBulkPartWriter}.
 ```
 
So we need this class only to modify the objects being returned from mentioned functions. And the actual workaround is described and implemented in `DeltaBulkPartWriter`.

Another thing is why we need it at all but to understand this one must get familiar with classes belonging to the package `org.apache.flink.streaming.api.functions.sink.filesystem` and `org.apache.flink.core.fs` in Flink (eg. `flink-file-sink-common` module which we are utilizing in out implementation). Those packages/modules/classes are like a generic ones for interacting with file systems in Flink.
Long story short all of the interfaces are build around the concept of classes like `BucketWriter`, `InProgressFileWriter`, `BulkBucketWriter`,  `RecoverableWriter` and so on which are user-facing abstractions. Then during runtime Flink resolves the correct implementations to use like `org.apache.flink.runtime.fs.hdfs.HadoopRecoverableWriter` or `org.apache.flink.core.fs.local.LocalRecoverableWriter` but on a sink-implementation level it's a black box for us.

allisonport-db committed:
Fix Github actions to fail when an sbt test fails (#212)

A few issues fixed:
- Github actions wasn't reporting any failed tests unless the last test on the list failed, since it restarted `build/sbt` each time, and Github actions only considers the final exit code.
- Moved `compatibility/test` to only run on Scala 2.12 since it uses Delta 1.0.0
-  Fixed a few JavaDoc compilation errors that were merged in #196.
   - [QUESTION] Documentation in #196 uses `@implNote` Javadoc tag. I added it to the Javadoc compilation, but should we just change that doc instead?
   - [NOTE] Seems related to https://github.com/sbt/sbt/issues/875, but when there is a compilation error, all the other warnings are reported as errors. For example in the previous commit, it showed 18 `[error]` messages but also: 
```
[info] 2 errors
[info] 18 warnings
[error] (standalone/javaunidoc:doc) javadoc returned nonzero exit code
```
- Fixed the other tests that were failing

pkubit-g committed:
Extend java docs && apply code review remarks

scottsand-db committed:
DSW convert-to-delta example (#205)

* example WIP

* delete file made in error

* remove parquet files, verify commit, generate less parquet

* move script to resoures

* dealing w existing/non-existant directory

* edit comments

* python instructions

* removing personal file path

* addressing comments

* add back source vs target path

* fix doc

* use delta log for check instead

* small changes

Co-authored-by: allisonport-db <allison.portis@databricks.com>
Co-authored-by: allisonport-db <89107911+allisonport-db@users.noreply.github.com>

zsxwing committed:
Fix flaky build and add Hive tests back (#217)

I noticed the following output in each failed build.

```
Files to process must be specified, found 0.
Files to process must be specified, found 0.
Error: Process completed with exit code 255.
```

After digging more, I found the root cause of our flaky build is the same as https://github.com/etsy/sbt-checkstyle-plugin/issues/32

sbt-checkstyle calls Java checkstyle library which will [call System.exit when the style check fails](https://github.com/etsy/sbt-checkstyle-plugin/blob/58b5c5fc4fc2d45b56ea07a9ad858089e34853e7/src/main/scala/com/etsy/sbt/checkstyle/Checkstyle.scala#L47-L52). In order to avoid exiting SBT JVM process, sbt-checkstyle [uses its own NoExitSecurityManager to prevent exits from the JVM](https://github.com/etsy/sbt-checkstyle-plugin/blob/58b5c5fc4fc2d45b56ea07a9ad858089e34853e7/src/main/scala/com/etsy/sbt/checkstyle/Checkstyle.scala#L122-L141). However, the implementation is [not thread thread](https://github.com/etsy/sbt-checkstyle-plugin/issues/32#issuecomment-976206148), and cannot prevent from exiting the JVM when two concurrent threads are running the [noExit](https://github.com/etsy/sbt-checkstyle-plugin/blob/58b5c5fc4fc2d45b56ea07a9ad858089e34853e7/src/main/scala/com/etsy/sbt/checkstyle/Checkstyle.scala#L122-L141) method.

As most of our projects have zero Java files, Java checkstyle library will fail and call System.exit for most of our projects. Due to the above race condition, the SBT JVM may be killed when sbt-checkstyle's `noExit` method cannot prevent from exiting the JVM. `exit code 255` in our output matches the following code in Checkstyle:
- https://github.com/checkstyle/checkstyle/blob/a42a3e1733f3c7583177a3c96de65e6a2fd7671f/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L87
- https://github.com/checkstyle/checkstyle/blob/a42a3e1733f3c7583177a3c96de65e6a2fd7671f/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L184-L189
- https://github.com/checkstyle/checkstyle/blob/a42a3e1733f3c7583177a3c96de65e6a2fd7671f/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L810-L812

The fix is just disabling parallel execution. While it makes the build slower, the speed is still acceptable since most of time is spent on tests which cannot run in parallel anyway.

This PR also adds build cache, hive tests back and upgrade Ubuntu back to 20.04. The SBT heap size is increased as well as I hit SBT OOM occasionally during my investigation.

pkubit-g committed:
Correct java docs

pkubit-g committed:
Correct Java docs
